std/array_list.zig

```diff
@@ -193,27 +194,93 @@ pub fn AlignedArrayList(comptime T: type, comptime A: u29) type {
             return self.pop();
         }
 
+        // an iterator which allows you to remove the current element
+        // while iterating through it.
         pub const Iterator = struct {
-            list: *const Self,
-            // how many items have we returned
-            count: usize,
+            list: *Self,
+
+            // index of the next element to return
+            next_index: usize = 0,
+
+            // did we remove the element last returned by `next`?
+            removed: bool = false,
 
             pub fn next(it: *Iterator) ?T {
-                if (it.count >= it.list.len) return null;
-                const val = it.list.at(it.count);
-                it.count += 1;
+                // if we removed the previous element, then
+                // the next element replaced the current one,
+                // so we get the element at the same index a second time.
+                // the next element's index was incremented when we
+                // called `next` before, so we decrement it here, before we get the item.
+                // NOTE: volatile code! don't use a ternary if, or stack variable for new index here!
+                if (it.removed) {
```

Comment:
Avoid this branch by putting the decrement into the remove function.
Then this can be a non-branching `it.removed = false`.

Reply:
Interestingly, in debug mode at least, what you suggest is actually slower than the way it currently is.